### PR TITLE
Add `deterministic` option into convolution_2d and deconvolution_2d

### PR DESCRIPTION
**Pull request by zori**
@t-abe's https://github.com/pfnet/chainer/pull/1189 can't be reopened.
TODO (requested by @delta2323): Add test cases for this option where we call the function twice and compare results.

Some of convolution algorithm cuDNN provides are non-deterministic.
So outputs of current chainer convolution and deconvolution can be non-deterministic when cuDNN is available.
I think this might make trouble when you want to do reproducible experiments.

This PR add deterministic option into these two functions and links.
If True, the functions force cuDNN to use deterministic algorithms.

### CODE REVIEW

File: chainer/functions/connection/deconvolution_2d.py [hunk 1]
```diff
@@ -291,6 +300,10 @@ def deconvolution_2d(x, W, b=None, stride=1, pad=0,
             input size, stride and pad.
         use_cudnn (bool): If ``True``, then this function uses cuDNN if
             available.
+        deterministic (bool): The output of this function can be
+            non-deterministic when it uses cuDNN.
+            If this option is `True`, then it forces cuDNN to use
```
Comment (unnonouno) [inline chainer/functions/connection/deconvolution_2d.py]: Use double backquotes.

File: tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py [hunk 2]
```diff
@@ -189,4 +189,92 @@ def test_call_cudnn_backrward(self):
...
+        """
+        x1, W1, b1, y1 = self._run()
+        x2, W2, b2, y2 = self._run()
+
+        cuda.cupy.testing.assert_array_equal(y1.data, y2.data)
+        if self.deterministic:
+            cuda.cupy.testing.assert_array_equal(W1.grad, W2.grad)
+        else:
+            with self.assertRaises(AssertionError):
+                cuda.cupy.testing.assert_array_equal(W1.grad, W2.grad)
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py]: Are these arrays always different when `deterministic=False`?

Comment (zori) [inline tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py]: Yes, for `numpy.float32` data types, cudnn's chosen backward filter algorithm is non-deterministic (CUDNN_CONVOLUTION_BWD_FILTER_ALGO_3) and the computed gradients of the weights in the two runs will be slightly different.

**Comment by zori** [comment]
I have noticed my test code does not use `_make_gpu_args()`, I will remove it, sorry.

File: tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py (hunk 2 above)
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py]: Is it possible to ensure results are different? I felt we don't need to ensure it. This test case is not required, is it?

Comment (zori) [inline tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py]: Do you mean the `else`-part is not necessary? It ensures that without the `deterministic` option, the  cudnn-chosen algorithm will produce different weight gradients. So it's like a test which proves the motivation to make this PR. However, it does not test the `deterministic` parameter. Should I modify or remove it?

Comment (unnonouno) [inline tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py]: How about checking if `getConvolutionBackwardFilterAlgorithm` is actually called using `mock` module?

**Comment by unnonouno** [comment]
Please check if py2 can pass the test.

**Comment by zori** [comment]
During development I tested with python 2.7.6. I just ran the nosetests for test_{de,}convolution_2d.py on python 3.5.2 (both scenarios passed).

**Comment by unnonouno** [comment]
I found that cuDNN v2 cannot support a deterministic algorithm because it doesn't have an interface to select an algorithm (it was introduced from v3). https://github.com/pfnet/chainer/pull/1321/files#diff-34a15eecaef34d920b685d1f971f6a17R219

We need to consider behavior when a user is using v2 and deterministic option is selected.

**Comment by unnonouno** [comment]
In my opinion, when a user uses cuDNN v2 and selected deterministic, we can raise an error.

**Comment by zori** [comment]
I added a note in the documentation about cudnn version that supports `deterministic` and raise an exception if user has lower version but selects `deterministic=True`. 

The test is changed to only check behavior when `deterministic=True` as the other tests cover the rest of the functionality of (de)convolution2D.

**Comment by unnonouno** [comment]
It looks good. Can you rebase or merge the master branch?

**Comment by zori** [comment]
Rebased. Thank you very much for the review.

**Comment by unnonouno** [comment]
Great!! Please check the result of travis.

**Comment by unnonouno** [comment]
Oh, it looks your change is removed.

**Comment by zori** [comment]
Sorry, the very first commit (c9cf1325cf01182e654f8c726a03e77f8e5a85e5) had already been merged into 'master', that's why it had disappeared after re-basing. Fixed and ran in docker.

**Comment by unnonouno** [comment]
Thank you! I'll try to test.

**Comment by unnonouno** [comment]
LGTM! Thank you very much.

**Pull request merged**